[Metrics] Model FLOPs Utilization estimation

SungMinCho opened the pull request:
Implements #24190

Summary:
Signed-off-by: Sung Min Cho sungmincho@meta.com

Use D82721353 to record average TF/s/GPU and GB/s/GPU periodically along with other existing stats like token throughput.

---

### Example outputs

common setup:
- model: gpt-oss
- batch size: 64
- input length: 6100
- output length: 1000

- TP8
  - full server log P2043309463
  - logger grep P2043310377
- TP4
  - full server log P2043310682
  - logger grep P2043310909
- TP4PP2
  - full server log P2043311278
  - logger grep P2043311447
- DP2TP4 - TP8
  - full server log P2043311757
  - logger grep P2043312147
- DP2TP4 - EP8
  - full server log P2043312361
  - logger grep P2043312522

Test Plan: CI

Differential Revision: D82742756

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,898 @@
+        args.hidden_size = get_required(cfg, "hidden_size")
+        args.num_attention_heads = get_required(cfg, "num_attention_heads")
+        args.num_hidden_layers = get_required(cfg, "num_hidden_layers")
+
```
In vLLM, we have some more general way to get these model parameters: https://github.com/vllm-project/vllm/blob/5f3cd7f7f20a8e4445d70cbd1f5475175ef391e3/vllm/config/model.py#L1230-L1353

In the future, we are moving towards a more unified config type to make this parsing easier: https://github.com/vllm-project/vllm/issues/24384

zhuohan123 reviewed (commented):
Thanks for the work! The integration is pretty standalone so it looks pretty nice to me. The biggest comment is that vLLM has some more general functions to get some of the parameters you want. We can make use of these functions to make the estimation more general.

zhuohan123 commented on vllm/v1/metrics/loggers.py:
```diff
@@ -268,6 +311,18 @@ def __init__(
+        # Enable only when DP==1 for multiple reasons
+        # - Currently, LoggingStatLogger will not get instantiated and only
+        #   AggregatedLoggingStatLogger will be used in case of DP==1 and
+        #   if entered from FB-entrypoint.
```
What's `FB-entrypoint`? Let's try to avoid internal-only logic here.

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+
+    num_tokens: int
+    context_len: int
+    is_prefill: bool
```
Why do we need to distinguish prefill and decode? From flops perspective they should be the same right?

SungMinCho replied:
For example, the read_bytes() will differ depending on whether it is prefill or decode (i.e. decode will "read" context_len amount of KV cache, whose precision might differ from activation_byte_size, and so on). So I think it's good to keep them separate.

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+        args.weight_byte_size = get_dtype_size(torch_dtype)
+
+        # FIXME: handle this better
+        args.activation_byte_size = 2
```
Can you leave a more detailed TODO for the hardcode here and how exactly should we fix it?

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+            args.weight_byte_size = 1
+        else:
+            # FIXME: Add more parsing logic for different quant methods.
+            pass
```
Should we error out here?

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+        else:
+            # FIXME: Add more parsing logic for different quant methods.
+            pass
+
```
ditto, should we error out here?

SungMinCho replied:
Updated to throw error.

zhuohan123 commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+            pass
+        elif quant_method == "mxfp4":
+            # FIXME: Also has "ignored layers" issue above
+            args.weight_byte_size = 0.5
```
Do we also need to count the block scale's size?

markmc commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+            try:
+                metric = metric_cls.from_vllm_config(vllm_config)
+                self.metrics.append(metric)
+            except ValueError:
```
I'm wondering whether debug logging would be valuable here - e.g. if you were extending or adding a component, this would currently just swallow a useful error

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -188,6 +189,15 @@ def __init__(
+        # Used to derive model's MFU stats for each SchedulerOutput
+        self.perf_metrics: ModelMetrics | None
+        try:
+            self.perf_metrics = ModelMetrics(vllm_config)
```
We should only enable this if `log_stats` is `True`

SungMinCho replied:
Done.

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -229,6 +262,16 @@ def log(self):
             log_parts.append("External prefix cache hit rate: %.1f%%")
             log_args.append(self.connector_prefix_caching_metrics.hit_rate * 100)
+
+        if self.has_perf_stats and self._enable_perf_stats():
```
Can't we just make this conditional on `last_avg_tflops_per_gpu` and `last_avg_gbps_per_gpu` being non-zero?

SungMinCho replied:
Done, but keeping _enable_perf_stats() for reasons mentioned in https://github.com/vllm-project/vllm/pull/28859#discussion_r2557955671

markmc reviewed (changes requested):
Some comments on a first pass through the implementation

SungMinCho commented on vllm/v1/metrics/loggers.py:
```diff
@@ -111,6 +112,12 @@ def __init__(self, vllm_config: VllmConfig, engine_index: int = 0):
+        self.last_avg_tflops_per_gpu = 0.0
+        self.last_avg_gbps_per_gpu = 0.0
+        self.tp_size = vllm_config.parallel_config.tensor_parallel_size
+        self.pp_size = vllm_config.parallel_config.pipeline_parallel_size
```
Removed tp_size which was not being used. Kept pp_size because it is used to divide duration.

SungMinCho commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,939 @@
+        return ParserChain(
+            BaseConfigParser(),
+            BaseAttentionConfigParser(),
+            AttentionQuantizationConfigParser(),
```
I see what you're saying and my initial approach was actually what you suggested but I found many duplicate code involving basic config parsing (repeated for each component) which was uglier than this... so I switched to composable and stack-able chain of parsers. I think it helps with extensibility. Same reason why I keep a list of ComponentMetrics and not a monolithic ModelMetrics that's responsible for everything. I guess I'm implicitly nudging people to add parsers/metrics/etc that they are interested in to iteratively make the whole thing more accurate in a gradual manner.

markmc commented on vllm/v1/metrics/perf.py:
```diff
@@ -0,0 +1,962 @@
+    def from_vllm_config(cls, vllm_config: VllmConfig) -> Self:
+        """
+        Instantiate this class from VllmConfig.
+        Raises ValidationError if parsing fails.
```
Raises `InvalidComponentError`

markmc commented:
I've hacked on this a bit and sent you some PRs:

- SungMinCho/vllm#1
- SungMinCho/vllm#2
- SungMinCho/vllm#3

markmc commented:
>  (2) merge @markmc 's PRs onto this PR

Treat my PRs as changes I request through code review - I'm just making it easier to communicate the changes I'd like to see and easier for you to incorporate those changes

SungMinCho commented:
Hi @markmc I included the first two of your PRs as a part of this PR and pushed just now. (I ask that we make the third prometheus PR as a separate follow up PR from you because I'm not really familiar with prometheus and don't feel too comfortable including it in this PR without being able to thoroughly test it myself). Please take a look.

I also dug deeper into the performance degradation issue. I found some inefficiencies and removed them. For example, using pydantic just as syntactic sugar for Breakdown() turned out to be very slow so I replaced them with plain dictionaries. **But the perf degradation still persists**. Please compare the following logs:
- with MFU: https://gist.github.com/SungMinCho/fe6bb9e1a67a94e4e46b57a92fdb8bc3
- without MFU: https://gist.github.com/SungMinCho/1265fae60f4e78af21cb9d572efeb2a7

You can see that without MFU, avg generation throughput hovers around 6000 tokens/s whereas with MFU it is brought down to around 5700 tokens/s. If you check the end of the lines you will see that "mfu calc overhead" is around 5%, which means total 5% of the interval was spent on calculating the MFU stats... I guess pure arithmetic **can** accumulate to non-trivial overhead when there are many requests to calculate the stats for and when the actual GPU operation is very fast in contrast (e.g. decode)... (Still, it only makes half-sense to me so I'm continuing to debug).

I tried some optimizations like lazily calculating the MFU stats by only accumulating `num_tokens`, `context_len` (and their product) across all the observed requests, and then running these sums through our flops/bytes equations **only once** at the last moment to calculate the total MFU stats for that logging interval, and I did see that the performance degradation was mitigated by a lot (overhead decreases from 5% -> 0.x%?), but then later I realized that this leads to incorrect MFU stats (e.g. we can't naively accumulate num_tokens and derive the sum of number of activated experts because the relationship is nonlinear)...

I'll keep trying to optimize as 5% perf degradation will definitely not be approved. Just wanted to share the progress.

SungMinCho committed:
analytic MFU stats (#28859)

Summary:

Signed-off-by: Sungmin Cho <sungmincho@meta.com>

Use D82721353 to record average TF/s/GPU and GB/s/GPU periodically along with other existing stats like token throughput.

---

### Example outputs

common setup:
- model: gpt-oss
- batch size: 64
- input length: 6100
- output length: 1000

- TP8
  - full server log P2043309463
  - logger grep P2043310377
- TP4
  - full server log P2043310682
  - logger grep P2043310909
- TP4PP2
  - full server log P2043311278
  - logger grep P2043311447
- DP2TP4 - TP8
  - full server log P2043311757
  - logger grep P2043312147
- DP2TP4 - EP8
  - full server log P2043312361
  - logger grep P2043312522

Test Plan: CI

Differential Revision: D82742756

SungMinCho committed:
config-based flops/bytes estimator

Summary:
Signed-off-by: Sungmin Cho <sungmincho@meta.com>

The purpose of the diff is to provide config-based performance metric estimator for vLLM.

In other words, we implement `ModelMetrics` class, which is instantiated with `VllmConfig` and outputs performance metric estimates (i.e. flops/bytes) for each given `ExecutionContext`.

The design goal is to make it easy for maintainers to
- add more components (by adding `ComponentMetrics` subclass)
- add more `VllmConfig` parsing logic (by adding `Parser` to the `ParserChain` associated with the target `ComponentMetrics`

So far we've only implemented Attn/Ffn/Unembed component metrics with minimum parsing logic to support couple of well-known configs such as gpt-oss, deepseek, llama4, etc.

Test Plan:
```
buck2 test //vllm/trunk:test_perf_metrics
```

Differential Revision: D82721353

SungMinCho commented:
Moved to https://github.com/vllm-project/vllm/pull/30737

Closing this.
